Update init-maven script to parse booster version numbers

amisevsk opened the pull request:
This is the first step for https://github.com/redhat-developer/rh-che/issues/670. 

- Extract common maven dependency downloading code into a helper script
- Parse version number from booster-catalog rest api [1]
- Add docker build script that copies helper script into subdirectories

One issue with the current approach is that the `build.include` script is included in the build context even for builds that do not need to parse boosters, but I think the effect is small enough that it's worth having cleaner code (and not handling booster/non-booster images separately).

[1] - https://forge.api.openshift.io/api/booster-catalog

amisevsk commented:
Also important note: the API does not surface a version for the wildfly-swarm configmap booster as there is not one available on OSIO, so that step is currently commented out. @ibuziuk should I create an issue to track this?

ibuziuk commented:
@amisevsk yeah, please would be nice to have the issue about missing wildfly cm in rh-che ?

amisevsk commented:
@l0rd @ibuziuk I've updated the PR following our discussion the other day. Now all shared parts are stored in the `context` directory. I've updated the readme to reflect the changes and provide some info about how to build locally / what the scripts do.

Let me know if there are any improvements I can make or if I missed anything.

amisevsk commented:
Some notes about the latest commit:
- I wasn't 100% sure on how to handle the step
  ```
  RUN sudo chgrp -R 0 /home/user \
    && sudo chmod -R g+rwX /home/user \
    && sudo chgrp -R 0 /etc/passwd \
    && sudo chmod -R g+rwX /etc/passwd \
    && sudo chgrp -R 0 /etc/group \
    && sudo chmod -R g+rwX /etc/group \
    && sudo mkdir -p /projects \
    && sudo chgrp -R 0 /projects \
    && sudo chmod -R g+rwX /projects \
    && sudo chgrp -R 0 /tmp \
    && sudo chmod -R g+rwX /tmp
  ```
  so it is executed along with the other default setup steps in dockerfiles (previously, it had been done both early or at the end of the file, depending on image)
- Also with the above, some images change group and permissions on the `/tmp` directory. I'm not clear on why this is necessary so those lines are commented out of the associated script. AFAIK any user can write to /tmp, and some of the earlier steps in the current master erase everything in the `/tmp` directory anyways.
- You'll notice that the wildfly-swarm dockerfile has the `configmap` booster commented out. This is because there is no supported version of this booster on `osio`. Once it is available this line should be uncommented (should we create an issue to track this?)

amisevsk commented:
@ibuziuk Sorry I missed your comment earlier. I have created https://github.com/redhat-developer/rh-che/issues/734 to track the wildfly-swarm configmap version info.

ibuziuk commented:
@amisevsk have you got a chance to test new images on osio / minishift ?

ibuziuk commented on README.md:
```diff
@@ -2,3 +2,27 @@
+docker build \
+  -t spring-boot:local \
+  -f ./recipes/spring-boot/Dockerfile \
+  ./context
```
just figured out with @amisevsk that building that way is quite a new (18.03 )feature [1] which would not be supported by automated build ci

[1] https://github.com/docker/cli/pull/886

amisevsk replied:
My bad I thought this was a sane enough feature that it would've been around since the `-f` option was included.

amisevsk commented on recipes/centos-nodejs/Dockerfile:
```diff
@@ -13,48 +13,26 @@ ARG OC_VERSION=3.6.173.0.30
-    sudo yum clean all && \
-    sudo rm -rf /tmp/* /var/cache/yum
+# Install general dependencies
+COPY --chown=user:user ["./general-deps.install", "/tmp/install/"]
```
Also the `--chown` option was added in docker 17.06

amisevsk commented:
@ibuziuk After our discussion today, I've updated the PR to be buildable on `docker 1.13.1-63` (confirmed by building in a centos7 VM). I'm now moving onto testing the images manually on OpenShift.

Changes: 
- remove `--chown` flag from `COPY` commands and instead add `chown` step to `RUN`. 
- reorganize repo so that context dir is in build path.

l0rd commented on README.md:
```diff
@@ -2,3 +2,27 @@
+docker build \
+  -t spring-boot:local \
+  -f ./recipes/dockerfiles/spring-boot/Dockerfile \
+  ./recipes/
```
Can't you set the build context to `./recipes/context`?

l0rd replied:
Got it. I didn't saw previous comments. Sorry about that.

ibuziuk commented:
@amisevsk for some reason when I try to build `spring-boot` via:

` docker build   -t spring-boot:local   -f ./recipes/dockerfiles/spring-boot/Dockerfile   ./recipes/
`

I keep getting:

```
Step 8/21 : RUN sudo chown user:user -R /tmp/install &&     /tmp/install/nodejs.install &&     sudo rm -rf /tmp/install
 ---> Running in 3c0f4597fac1
/bin/sh: /tmp/install/nodejs.install: /bin/bash: bad interpreter: Text file busy
The command '/bin/sh -c sudo chown user:user -R /tmp/install &&     /tmp/install/nodejs.install &&     sudo rm -rf /tmp/install' returned a non-zero code: 126
```

amisevsk commented:
@ibuziuk That is especially strange, as the failure is due to a command being run *inside* the container failing -- I don't really understand how this could happen since it should be isolated from everything on your system. 

The only thing I can imagine having the file open is the `chown` command, which should have completed. Does it still happen if you change the previous command to
```
RUN sudo chown user:user /tmp/install tmp/install/* && [...]
```
?

ibuziuk commented:
@amisevsk removing recursive option helped but than it failed with the following:

```
[INFO] Scanning for projects...
Downloading: https://repo1.maven.org/maven2/io/openshift/booster/spring-boot-booster-parent/1.5.12-3-rhoar/spring-boot-booster-parent-1.5.12-3-rhoar.pom
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for io.openshift.booster:spring-boot-configmap:1.5.12-1-osio: Could not transfer artifact io.openshift.booster:spring-boot-booster-parent:pom:1.5.12-3-rhoar from/to central (https://repo1.maven.org/maven2): GET request of: io/openshift/booster/spring-boot-booster-parent/1.5.12-3-rhoar/spring-boot-booster-parent-1.5.12-3-rhoar.pom from central failed and 'parent.relativePath' points at wrong local POM @ line 20, column 11
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project io.openshift.booster:spring-boot-configmap:1.5.12-1-osio (/home/user/tmp-folder/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for io.openshift.booster:spring-boot-configmap:1.5.12-1-osio: Could not transfer artifact io.openshift.booster:spring-boot-booster-parent:pom:1.5.12-3-rhoar from/to central (https://repo1.maven.org/maven2): GET request of: io/openshift/booster/spring-boot-booster-parent/1.5.12-3-rhoar/spring-boot-booster-parent-1.5.12-3-rhoar.pom from central failed and 'parent.relativePath' points at wrong local POM @ line 20, column 11: Tag mismatch! -> [Help 2]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
The command '/bin/sh -c sudo chown user:user -R /tmp/maven-deps &&     /tmp/maven-deps/maven-deps.install         'spring-boot'         'configmap'         'health-check'         'rest-http' &&     sudo rm -rf /tmp/maven-deps' returned a non-zero code: 1
```

amisevsk commented:
@ibuziuk I am at a loss on that issue -- I don't know why the maven build would fail in your containers when it works fine in mine (on docker 1.8 and 1.3). @l0rd Any insight?

l0rd commented:
`docker build -t spring-boot:local -f ./recipes/dockerfiles/spring-boot/Dockerfile ./recipes/` works fine for me. No errors. I have tested on fedora with docker version 1.13.1 and macos with docker for mac 18.03.1.

ibuziuk commented:
@l0rd @amisevsk I'm ok to merge if it works for you guys and on centos.ci! (I will try to figure out myself why it is failing on my machine)

amisevsk commented:
I'd prefer to at least quickly test the images on OSIO before merging, but I cannot create workspaces. The images are pushed to `amisevsk/[image]:pr677` (I tagged them wrong).

ibuziuk commented:
@amisevsk what about creating a duplicate of stack on che.opeshift.io and update the image in configuration ?

amisevsk commented:
After further testing I found that attempting to run a maven build in the images created by this PR fails due to 
```
[WARNING] Failed to create parent directories for tracking file /home/user/.m2/repository/io/openshift/booster-parent/23/booster-parent-23.pom.lastUpdated
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for io.openshift.booster:spring-boot-health-check:1.5.14-1-SNAPSHOT: Could not transfer artifact io.openshift:booster-parent:pom:23 from/to central (https://repo1.maven.org/maven2): /home/user/.m2/repository/io/openshift/booster-parent/23/booster-parent-23.pom.part.lock (No such file or directory) and 'parent.relativePath' points at wrong local POM @ line 19, column 11
```
After investigating this, it appears to be due to the PR moving the `grant-access-arbitrary-UID.sh` RUN command earlier in some docker files. The subsequent maven build command would create directories in the `.m2` folder with incorrect permissions. I've pushed another commit that moves this command to the end of the dockerfiles and it appears to have fixed the issue on my end. 

The directory structures (with permissions) are
```
[current OSIO image]
drwxrwxr-x  4 1000 root 4096 Jul 13 05:41 .m2
drwxrwxr-x 37 1000 root 4096 Jul 13 05:45 \_ repository
drwxrwxr-x. 1 1000 root   43 Jul 13 05:58   \_ io
drwxrwxr-x. 1 1000 root   28 Jul 12 23:16     \_ openshift
drwxrwxr-x. 1 1000 root   40 Jul 12 23:16       \_ booster
drwxrwxr-x. 1 1000 root   16 Jul 13 05:57       \_ booster-parent
drwxrwxr-x. 1 1000 root   97 Jul 12 23:16         \_ 21
drwxr-xr-x. 2 user root   97 Jul 13 05:57         \_ 23

[PR before last commit]
drwxrwxr-x  5 1000 root 4096 Jun 19 22:35 .m2
drwxrwxr-x 69 1000 root 4096 Jul  5 17:27  \_ repository
drwxr-xr-x  5 1000 user 4096 Jul  5 17:26    \_ io
drwxr-xr-x  4 1000 user 4096 Jul  5 17:25      \_ openshift
drwxr-xr-x  3 1000 user 4096 Jul  5 17:25        \_ booster
drwxr-xr-x  4 1000 user 4096 Jul  5 17:27        \_ booster-parent
drwxr-xr-x  2 1000 user 4096 Jul  5 17:25          \_ 21
drwxr-xr-x  2 1000 user 4096 Jul  5 17:27          \_ 22

[PR with latest commit]
drwxrwxr-x  6 1000 root 4096 Jul 13 06:05 .m2
drwxrwxr-x 78 1000 root 4096 Jul 13 06:05 \_ repository
drwxrwxr-x  9 1000 root 4096 Jul 13 06:05   \_ io
drwxrwxr-x  7 1000 root 4096 Jul 13 06:05     \_ openshift
drwxrwxr-x  4 1000 root 4096 Jul 13 04:50       \_ booster
drwxrwxr-x  7 1000 root 4096 Jul 13 06:05       \_ booster-parent
drwxrwxr-x  2 1000 root 4096 Jul 13 04:50         \_ 21
drwxrwxr-x  2 1000 root 4096 Jul 13 04:50         \_ 22
drwxr-xr-x  2 user root 4096 Jul 13 06:05         \_ 23
```
One detail I'm still not fully clear on is where the `23` directory comes from, as it has different permissions than the rest.

@ibuziuk when you have a moment could you check if this resolves the issue you were having with building the images?

ibuziuk commented:
@amisevsk unfortunately I still can not build locally :-(

```
Step 21/21 : RUN sudo chown user:user -R /tmp/install &&     /tmp/install/grant-access-arbitrary-UID.sh &&     sudo rm -rf /tmp/install
 ---> Running in 97da77b5a174
/bin/sh: /tmp/install/grant-access-arbitrary-UID.sh: /bin/bash: bad interpreter: Text file busy
The command '/bin/sh -c sudo chown user:user -R /tmp/install &&     /tmp/install/grant-access-arbitrary-UID.sh &&     sudo rm -rf /tmp/install' returned a non-zero code: 126

```

ibuziuk commented:
https://github.com/ibuziuk/che-dockerfiles-1/commit/b37e5300e09d16f1f2d66084612a7cc30e31299c
Removing recursive option fixes the build for me. Once `-R` was removed I was able to build the image & booster locally

amisevsk commented:
@ibuziuk Alright, I will make that change just to be safe.

ibuziuk commented:
@amisevsk I'm not sure it is required - if it builds via CI & works on osio than everything is fine - no need to remove `-R`, since it looks like a bug in my docker version

amisevsk commented:
@ibuziuk The commands *should* be identical but I'd rather not risk it (and it helps that this means you can still build the images locally). I haven't tested on CI directly, but I set up a VM with the same env and it worked there -- there's still a chance the `-R` option could cause issues down the road.

ibuziuk commented:
@amisevsk I have tested on prod against `ibuziuk/vertx:rh-che-670` build from this PR with `vertx-http-booster` and all seems to work fine. :+1: (build / run / debug)

+1 to merge, but extra checking new recipes much appreciated

[1] https://hub.docker.com/r/ibuziuk/vertx/tags/

amisevsk commented:
I've updated the PR to fix a few issues that came up in more thorough testing:
1. dotnet image failed to launch. Fixed by installing C# lang server as in https://github.com/eclipse/che-dockerfiles/blob/47505fa5026cfeb80da8d48328e77e1351b0646d/recipes/dotnet_core/Dockerfile#L25
2. nodejs stack doesn't include node. Fixed by modifying `nss_wrapper` install command, as it was uninstalling make, which node has as a dependency.
   - This one is a bit of a hack, I'd rather not have separate scripts to do the same thing. We should consider/look into using the nss_wrapper package included in the EPEL repo.

Related issues: https://github.com/redhat-developer/rh-che/issues/763

amisevsk commented:
I've done a quick standard use-case test of all of these images now, and as far as I can see, the only outstanding issue in the PR is https://github.com/redhat-developer/rh-che/issues/764, which is outside the scope of this PR.

rhopp commented:
.NET image: Workspace startup is successful with the latest commit, but there are lot of error markers in sourcecode:
![image](https://user-images.githubusercontent.com/1215011/42939914-cf76baaa-8b57-11e8-96dd-69d9aaf0c07d.png)
Building and running works fine, with one minor drawback - The `preview` link is not resolved correctly:
![image](https://user-images.githubusercontent.com/1215011/42939958-f3b44158-8b57-11e8-8715-cec3eaa0a6bf.png)
One has to go to `servers` and take the preview URL from there.

ibuziuk commented:
@rhopp could you please add it to https://github.com/redhat-developer/rh-che/issues/763 ? I think it should be handled outside the scope of this PR

rhopp commented:
@ibuziuk Sure. Agree that should be out of scope of this one... I just wasn't aware of that issue.

amisevsk commented:
Pushed another commit fixing the `npm` issue @rhopp described. In the centos-nodejs image, gcc is already installed and uninstalling it removes npm as a dependant.

ibuziuk commented:
@rhopp @amisevsk are we good to merge now ? was absence of `npm` the only spotted issue in comparison with images used on osio ?
